mailpoet/tests/unit/Subscribers/SubscriberSubscribeControllerUnitTest.php

```diff
@@ -0,0 +1,494 @@
+    ]);
+    $captchaSession = Stub::makeEmpty(CaptchaSession::class,
+      [
+        'getCaptchaHash' => $captcha,
+        'init' => function($receivedSessionId) use ($captchaSessionId) {
+          expect($receivedSessionId)->equals($captchaSessionId);
+        },
+      ]);
+    $formFields = [
+      'field_a' => 'value_a',
+      'field_b' => 'value_b',
+    ];
+    $submitData = array_merge([
+      'captcha_session_id' => $captchaSessionId,
+      'captcha' => $captcha,
+    ], $formFields);
+    $segmentIds = [1];
+    $form = Stub::makeEmpty(
+      FormEntity::class,
+      [
+        'getSettingsSegmentIds' => function() use ($segmentIds): array {
+          return $segmentIds;
+        },
+        'getBlocksByTypes' => function() use ($formFields) {
+          $fields = array_values(array_map(
+            function(string $id): array {
+              return [
+                'id' => $id,
+              ];
+            }, array_keys($formFields)
+          ));
+
+          return $fields;
+        },
+      ]
+    );
+    $subscriber = Stub::makeEmpty(SubscriberEntity::class);
+    $subscriberActions = Stub::make(
+      SubscriberActions::class,
+      [
+        'subscribe' => function($receivedData, $receivedSegmentIds) use ($formFields, $segmentIds, $subscriber) {
+
+          expect($receivedData)->equals($formFields);
+          expect($receivedSegmentIds)->equals($segmentIds);
+          return $subscriber;
+        },
+      ],
+      $this
+    );
+    $subscriptionUrlFactory = Stub::makeEmpty(SubscriptionUrlFactory::class,
+      [
+        'subscribe' => function($receivedSubscriber, $receivedForm) use ($subscriber, $form) {
+          expect($receivedSubscriber)->equals($subscriber);
+          expect($receivedForm)->equals($form);
+        },
+      ]);
+    $throttling = Stub::makeEmpty(SubscriptionThrottling::class);
+    $fieldNameObfuscator = Stub::makeEmpty(FieldNameObfuscator::class,
+    [
+      'deobfuscateFormPayload' => function($data) { return $data;
+      },
+    ]);
+    $requiredCustomFieldValidator = Stub::makeEmpty(RequiredCustomFieldValidator::class);
+    $settings = Stub::makeEmpty(
+      SettingsController::class,
+      [
+        'get' => function($value) {
+          if ($value === 'captcha') {
+            return [
+              'type' => Captcha::TYPE_BUILTIN,
+            ];
+          }
+        },
+      ]
+    );
+    $formsRepository = Stub::makeEmpty(
+      FormsRepository::class,
+      [
+        'findOneById' => function() use ($form): FormEntity {
+          return $form;
+        },
+      ]
+    );
+    $statisticsFormsRepository = Stub::makeEmpty(StatisticsFormsRepository::class);
+    $wp = Stub::make(
+      WPFunctions::class,
+    [
+      'doAction' => function($receivedHook, $receivedData, $receivedSegmentIds, $receivedForm) use ($formFields, $segmentIds, $form) {
+        expect($receivedHook)->equals('mailpoet_subscription_before_subscribe');
```

Comment:
It's great that you have added tests! One more idea is to test how throwing an exception from this hook interrupts the subscription process and returns the exception message. That would serve as a test for the spec 2:
> A message from the exception thrown from the filter should be displayed as an error within a form to a user who wants to subscribe

Reply:
7d345c5 Adds an unit test to make sure the subscription does not happen.
24d3808 Adds an integration test checking the same and whether the API output is expected.